crabml-core/src/backends/cpu/primitives/matmul_vec.rs

```diff
@@ -65,20 +65,31 @@ fn gemv_simd<'a>(
 
     let bufc = bufc.as_f32_mut();
     let bufb = {
-        let _t = metrics.matmul_quantize_walltime.track();
+        // let _t = metrics.matmul_quantize_walltime.track();
         &bufb.quantize(bufa.vec_dot_rhs_dtype()).unwrap()
     };
 
-    let _t = metrics.matmul_vec_dot_walltime.track();
+    // let _t = metrics.matmul_vec_dot_walltime.track();
 
     let k = bufb.len();
-    bufc.par_chunks_exact_mut(4)
-        .enumerate()
-        .for_each(|(cn, cp)| {
-            let mi = cn * 4;
-            cp[0] = bufa.vec_dot(mi * k, bufb, 0, k);
-            cp[1] = bufa.vec_dot((mi + 1) * k, bufb, 0, k);
-            cp[2] = bufa.vec_dot((mi + 2) * k, bufb, 0, k);
-            cp[3] = bufa.vec_dot((mi + 3) * k, bufb, 0, k);
-        });
+    // Just a observation, if the buffer is small, it's faster to use single thread
+    // And most of the time, the small buffer is less than 1_000
```

Comment:
how about try a benchmark over a bigger model like openllama 3B ? imho the bigger models ara highly likely to exceed the limit quq

Reply:
@MrCroxx can you help take a look? sorry i did not have a x86 environment at this moment

Reply:
it seems that the test is passing? what's the unexpected output it produces?

Reply:
> it seems that the test is passing? what's the unexpected output it produces?

It repeat one sentence like "She is bad. She is bad.She is bad..."

Reply:
which model? if it's tinyllamas and it's likely to go to these corner cases, the bigger models might behave better. you can try improve the temprature while generating tokens by passing a parameter like `-t 0.9`

Reply:
tinyllamas-stories-15m-q8_0 , I think it's a bug. Since it is not corner on my patch.

Reply:
> ```
> buc len: 288
> buc len: 288
> buc len: 288
> buc len: 288
> buc len: 768
> buc len: 768
> buc len: 288
> buc len: 32000
> ```
> 
> Seems a really big buffer comes after a lot small bufffers .
> 

288 is the embedding vector size of this model, i believe that the bigger models are more likely to have a longer embedding vector.

the 32000 one is multiplying the embedding table (a 32000 x 288 matrix), it contains 32000 tokens, while each token contains 288 dimensions.

Reply:
> tinyllamas-stories-15m-q8_0 , I think it's a bug. Since it is not corner on my patch.

never mind, if the unit test about tinyllamas did not break, it's not considered as broken.

we can run it over a bigger one like a 3B or 2B model and validate the output.

i found a x86 machine and it's output looks not bad to me:

```
$ cargo build --release && time  ./target/release/crabml-cli -m ./testdata/open-llama-3b-q8_0.gguf "captain america" --steps 100 -t 0 -T 2
warning: `/home/flaneur/.cargo/config` is deprecated in favor of `config.toml`
note: if you need to support cargo 1.38 or earlier, you can symlink `config` to `config.toml`
    Finished `release` profile [optimized] target(s) in 0.04s
loaded model: 90ms
captain america: the winter soldier
Captain America: The Winter Soldier (2014)
Captain America: The Winter Soldier (2014) is a movie starring Chris Evans, Scarlett Johans
```